lib/js/jsffi.nim

```diff
@@ -0,0 +1,382 @@
+#
+#
+#            Nim's Runtime Library
+#        (c) Copyright 2017 Nim Authors
+#
+#    See the file "copying.txt", included in this
+#    distribution, for details about the copyright.
+#
+
+## This Module implements types and macros to facilitate the wrapping of, and
+## interaction with JavaScript libraries. Using the provided types ``JsObject``
+## and ``JsAssoc`` together with the provided macros allows for smoother
+## interfacing with JavaScript, allowing for example quick and easy imports of
+## JavaScript variables:
+##
+## .. code-block:: nim
+##
+##  # Here, we are using jQuery for just a few calls and do not want to wrap the
+##  # whole library:
+##
+##  # import the document object and the console
+##  magicVar(document, JsObject)
+##  magicVar(console, JsObject)
+##  # import the "$" function
+##  proc jq(selector: JsObject): JsObject {. importcpp: "$(#)" .}
+##
+##  # Use jQuery to make the following code run, after the document is ready.
+##  # This uses an experimental ``.()`` operator for ``JsObject``, to emit
+##  # JavaScript calls, when no corresponding proc exists for ``JsObject``.
+##  proc main =
+##    jq(document).ready(proc() =
+##      console.log("Hello JavaScript!")
+##    )
+##
+
+when not defined(js) and not defined(nimdoc) and not defined(nimsuggest):
+  {.fatal: "Module jsFFI is designed to be used with the JavaScript backend.".}
+
+import macros
+
+const
+  setImpl = "#[#] = #"
+  setImplJsString = "#[toJSStr(#)] = #"
```

Comment:
`toJSStr` is not guaranteed to be present in the resulting code. For it to appear Nim has to know that it is used. What you should do instead is just change proc argument to `cstring`, and importcpp would be just `#[#] = #`.